### PR TITLE
DRAFT: Temporary PR to run tests on Travis CI

### PR DESCRIPTION
**Pull request by bender2k14**
The head of this branch is the first commit of the branch in PR #207.  It is the head of this branch (i.e. the first commit of the other branch) where tests are failing for me.  The second commit and head of the branch in PR #207 is merely the first workaround I found to get the tests to pass.  It might not be the fix we want.

### CODE REVIEW

**Comment by bender2k14** [comment]
The two tests that fail for me on my machine also fail in the Travis CI build.  One possible way to get these tests to pass is commit https://github.com/bender2k14/fsharp-hedgehog/commit/19a87608c61254060a86f6add27959e4f99cb5f8, which is the head of the branch in PR #207 .

**Comment by moodmosaic** [comment]
Yep, those two fail on my machine too now :confused: 

```
  X Hedgehog.Tests.MinimalTests.greedy traversal with a predicate yields the perfect minimal shrink [49ms]
  Error Message:
   System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at Microsoft.FSharp.Core.FSharpFunc`2.InvokeFast[V](FSharpFunc`2 func, T arg1, TResult arg2)
   at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 12
   at Hedgehog.Random.map@40-1.Invoke(Seed seed, Int32 size) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 41
   at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 12
   at Hedgehog.Random.run[a](Seed seed, Int32 size, Random`1 r) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 15
   at Hedgehog.Random.resize@80.Invoke(Seed seed, Int32 _arg1) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 81
   at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 12
   at Hedgehog.Random.run[a](Seed seed, Int32 size, Random`1 r) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 15
   at Hedgehog.Gen.run@45T.Invoke(Seed seed, Random`1 random) in ..\fsharp-hedgehog\src\Hedgehog\Gen.fs:line 46
   at Hedgehog.Gen.bindRandom@41.Invoke(Seed seed0, Int32 size) in ..\fsharp-hedgehog\src\Hedgehog\Gen.fs:line 48
   at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 12
   at Hedgehog.Random.delay@18.Invoke(Seed seed, Int32 size) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 19
   at Hedgehog.Random.unsafeRun[a](Seed seed, Int32 size, Random`1 _arg1) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 12
   at Hedgehog.Random.run[a](Seed seed, Int32 size, Random`1 r) in ..\fsharp-hedgehog\src\Hedgehog\Random.fs:line 15
   at Hedgehog.Property.loop@338-7.Invoke(Seed seed, Int32 size, Int32 tests, Int32 discards) in ..\fsharp-hedgehog\src\Hedgehog\Property.fs:line 348
   at Hedgehog.Property.Report(Int32 n, Property`1 p) in ..\fsharp-hedgehog\src\Hedgehog\Property.fs:line 361
   at Hedgehog.Property.Report(Property`1 p) in ..\fsharp-hedgehog\src\Hedgehog\Property.fs:line 365
   at Hedgehog.Property.Check(Property`1 p) in ..\fsharp-hedgehog\src\Hedgehog\Property.fs:line 376
   at Hedgehog.Tests.MinimalTests.greedy traversal with a predicate yields the perfect minimal shrink() in ..\fsharp-hedgehog\tests\Hedgehog.Tests\MinimalTests.fs:line 88
```

```
  X Hedgehog.Tests.MinimalTests.genExp is not its default value [47ms]
  Error Message:


MinimalTests.genExp |> MinimalTests.isDefaultGen |> not
Gen (Random null) |> MinimalTests.isDefaultGen |> not
true |> not
false

Expected: True
Actual:   False
  Stack Trace:
     at Hedgehog.Tests.MinimalTests.genExp is not its default value() in ..\fsharp-hedgehog\tests\Hedgehog.Tests\MinimalTests.fs:line 84
```

**Comment by moodmosaic** [comment]
Well, wow... :eyes:

**Comment by moodmosaic** [comment]
@bender2k14, one thought... Could those failing tests be an xUnit.net (runner) issue?

I tried taking out Paket *only* from `Hedgehog.fsproj` and got no failing tests, as per

```diff
$ git diff src/Hedgehog/Hedgehog.fsproj
diff --git a/src/Hedgehog/Hedgehog.fsproj b/src/Hedgehog/Hedgehog.fsproj
index edbf76f..e9f7282 100644
--- a/src/Hedgehog/Hedgehog.fsproj
+++ b/src/Hedgehog/Hedgehog.fsproj
@@ -16,9 +16,9 @@
     <Compile Include="Gen.fs" />
     <Compile Include="Property.fs" />
     <Compile Include="LinqSupport.fs" />
-    <None Include="paket.references" />
-    <None Include="paket.template" />
     <None Include="Script.fsx" />
   </ItemGroup>
-  <Import Project="..\..\.paket\Paket.Restore.targets" />
+  <ItemGroup>
+    <PackageReference Update="FSharp.Core" Version="4.1.17" />
+  </ItemGroup>
 </Project>
```

```
Test Run Successful.
Total tests: 185
     Passed: 185
 Total time: 5.9699 Seconds
     1>Done Building Project "Hedgehog.Tests.fsproj" (VSTest target(s)).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:01:33.60
```

File: tests/Hedgehog.Tests/Hedgehog.Tests.fsproj [hunk 1]
```diff
@@ -12,10 +12,21 @@
...
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="..\..\src\Hedgehog\Hedgehog.fsproj" />
   </ItemGroup>
-  <Import Project="..\..\.paket\Paket.Restore.targets" />
+  <ItemGroup>
+    <PackageReference Update="FSharp.Core" Version="4.2.3" />
```
Comment (moodmosaic) [inline tests/Hedgehog.Tests/Hedgehog.Tests.fsproj:30]: If I understand how `Update` and `Include` work, we want `Include` here, as we're specifying that version only to satisfy Unquote which requires version 4.2.3 of FSharp.Core

**Comment by moodmosaic** [comment]
I spend some time in the console trying to figure out why those two tests fail when we take out Paket but couldn't seem to find the reason :confused: I haven't touched a debugger for this yet though...

**Comment by bender2k14** [comment]
I am also working on this.  I am keeping your comments in mind even though I haven't been directly replying to them.

I am in the process of creating a MWE.  Here is the F# code that I hope will suffice for that.  Have yet to get minimal project/solution configuration to make the test pass or fail.

```F#
module Tests

open Xunit

type Container = Container of (unit -> unit)

// Warning FS0040
// This and other recursive references to the object(s) being defined will be checked for initialization-soundness at runtime through the use of a delayed reference. This is because you are defining one or more recursive objects, rather than recursive functions.
let rec container : Container =
    (fun _ -> match container with Container _ -> ()) |> Container

[<Fact>]
let ``container is not null``() =
    obj.ReferenceEquals(container, null) |> Assert.False
```

**Comment by bender2k14** [comment]
This draft PR has served its purpose.  I found a better work around in https://github.com/hedgehogqa/fsharp-hedgehog/pull/207#issuecomment-655761938.  Will close next time I am able.  (Can't from the phone app I am using now.)

**Comment by moodmosaic** [comment]
:+1: I can close it, if you want.

**Comment by bender2k14** [comment]
Ah, yea :D  I didn't think my app could, but then I found the feature tucked away.

**Pull request closed without merging**